qinspect/middleware.py

```diff
@@ -9,98 +9,132 @@
 except ImportError:
-    class MiddlewareMixin(object):
+
+    class MiddlewareMixin:
         def __init__(self, get_response=None):
             pass
 
 
 try:
-    from django.db.backends.utils import CursorDebugWrapper
+    from django.db.backends.utils import CursorWrapper, CursorDebugWrapper
 except ImportError:
-    from django.db.backends.util import CursorDebugWrapper
+    from django.db.backends.util import CursorWrapper, CursorDebugWrapper
 
-if hasattr(logging, 'NullHandler'):
+if hasattr(logging, "NullHandler"):
     NullHandler = logging.NullHandler
 else:
+
     class NullHandler(logging.Handler):
         def emit(self, record):
             pass
 
+
 log = logging.getLogger(__name__)
 log.addHandler(NullHandler())
 
 cfg = dict(
-    enabled=(settings.DEBUG and
-        getattr(settings, 'QUERY_INSPECT_ENABLED', False)),
-    log_stats=getattr(settings, 'QUERY_INSPECT_LOG_STATS', True),
-    header_stats=getattr(settings, 'QUERY_INSPECT_HEADER_STATS', True),
-    log_queries=getattr(settings, 'QUERY_INSPECT_LOG_QUERIES', False),
-    log_tbs=getattr(settings, 'QUERY_INSPECT_LOG_TRACEBACKS', False),
-    duplicate_min=getattr(settings, 'QUERY_INSPECT_DUPLICATE_MIN', 2),
-    roots=getattr(settings, 'QUERY_INSPECT_TRACEBACK_ROOTS', None),
-    stddev_limit=getattr(settings, 'QUERY_INSPECT_STANDARD_DEVIATION_LIMIT',
-        None),
-    absolute_limit=getattr(settings, 'QUERY_INSPECT_ABSOLUTE_LIMIT', None),
-    sql_log_limit=getattr(settings, 'QUERY_INSPECT_SQL_LOG_LIMIT', None),
+    enabled=getattr(settings, "QUERY_INSPECT_ENABLED", False),
+    log_stats=getattr(settings, "QUERY_INSPECT_LOG_STATS", True),
+    header_stats=getattr(settings, "QUERY_INSPECT_HEADER_STATS", True),
+    log_queries=getattr(settings, "QUERY_INSPECT_LOG_QUERIES", False),
+    log_tbs=getattr(settings, "QUERY_INSPECT_LOG_TRACEBACKS", False),
+    duplicate_min=getattr(settings, "QUERY_INSPECT_DUPLICATE_MIN", 2),
+    roots=getattr(settings, "QUERY_INSPECT_TRACEBACK_ROOTS", None),
+    stddev_limit=getattr(
+        settings, "QUERY_INSPECT_STANDARD_DEVIATION_LIMIT", None
+    ),
+    absolute_limit=getattr(settings, "QUERY_INSPECT_ABSOLUTE_LIMIT", None),
+    sql_log_limit=getattr(settings, "QUERY_INSPECT_SQL_LOG_LIMIT", None),
 )
 
-__all__ = ['QueryInspectMiddleware']
+__all__ = ["QueryInspectMiddleware"]
 
 _local = threading.local()
 
 
 class QueryInspectMiddleware(MiddlewareMixin):
-
-    class QueryInfo(object):
-        __slots__ = ('sql', 'time', 'tb')
-
-    sql_id_pattern = re.compile(r'=\s*\d+')
+    """A class to inspect Django SQL queries"""
+
+    class QueryInfo:
+        __slots__ = ("sql", "time", "tb", "summaries")
+
+    sql_id_pattern = re.compile(r"=\s*\d+")
+    registry = CollectorRegistry()
+
+    sql_query_latency = Gauge(
+        name="django_sql_query_latency",
+        documentation="The latency of django SQL query in milliseconds",
+        labelnames=["file", "linenum"],
```

Comment:
I'm ignorant about prometheus, and I'm going through their docs. They have a warning on labels

https://prometheus.io/docs/practices/naming/#labels

_CAUTION: Remember that every unique combination of key-value label pairs represents a new time series, which can dramatically increase the amount of data stored. Do not use labels to store dimensions with high cardinality (many different label values), such as user IDs, email addresses, or other unbounded sets of values._

If we arbitrarily label with filenames, does that arbitrarily grow the number of time series?

Reply:
I don't know if the number of labels will grow, since we don't know how many slow queries we have. Also, it depends on the latency limit in config (the lower is the limit, the higher is a number of unique slow queries)